### PR TITLE
feat: Flat Run Actions

### PR DESCRIPTION
**Pull request by keita-determined**
## Ticket

[ET-110]

## Description

Flat Run Actions
`Tensorboard` and `Pause/Unpause` actions are out of this PR, so not included
`Move` action will be implemented after #9390 is merged into this

## Test Plan

- Enable Run table
- Check if action button is visible when one or more flat runs are selected
- Check if `archive`, `unarchive`, `delete`, `kill` actions are working with/without RBAC
- If runs belong to searches, the warning modal should show up. if user cancel it, move action won't happen. 

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

[ET-110]: https://hpe-aiatscale.atlassian.net/browse/ET-110?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by keita-determined** [comment]
API doesnt not include `archived` value in FlatRun, so expose it if its not intentional

File: master/internal/api_runs.go
Comment (corban-beaird) [inline master/internal/api_runs.go:142]: @keita-determined 
question: It looks like else where we are also checking the archived status the experiment & project, are we wanting to do the same with how we retrive it in this case?
cc: @AmanuelAaron

Comment (AmanuelAaron) [reply master/internal/api_runs.go:142]: This should be `"r.archived OR e.archived"`

Comment (keita-determined) [reply master/internal/api_runs.go:142]: i found a potential bug.  
When i switch SQL with `r.archived OR e.archived AS archived`, the unarchive action does not unarchive experiment.archived. that causes this unexpected bug. this is not what we want, right?

https://github.com/determined-ai/determined/assets/109113616/fdb94e24-2ced-4921-afec-187ad08ca809

File: webui/react/src/pages/FlatRuns/FlatRunActionButton.tsx
Comment (keita-determined) [reply webui/react/src/pages/FlatRuns/FlatRunActionButton.tsx:43]: Yes

**Comment by keita-determined** [comment]
ready for review other than the unarchive action behavior. we have not decided what behavior we want for it, so probably make a ticket for it?

**Comment by ashtonG** [comment]
we need product approval before merging anyways, so we might be able to get a decision on the spot.

**Review by ashtonG — COMMENTED** [review]
we're targeting 80% test coverage with these prs but codecov is reporting 73 -- i think testing the move submission flow should get us over the line.

File: webui/react/src/pages/FlatRuns/FlatRunMoveModal.tsx
Comment (johnkim-det) [inline webui/react/src/pages/FlatRuns/FlatRunMoveModal.tsx:43]: prefer `() => void` since it still works with async but we may want non-async handlers too?

**Review by johnkim-det — COMMENTED** [review]
looks like modal initially displays project id instead of project name
Screenshot 2024-06-04 at 12 34 18 PM

File: webui/react/src/pages/FlatRuns/FlatRunMoveModal.tsx
Comment (keita-determined) [reply webui/react/src/pages/FlatRuns/FlatRunMoveModal.tsx:43]: i'll change it

File: webui/react/src/utils/flatRun.ts
Comment (johnkim-det) [inline webui/react/src/utils/flatRun.ts:9]: question: is there a reason we have `canDeleteFlatRun` and `canMoveFlatRun` but not `canCreateExperiment` and `canModifyExperiment`? seems like this could be consistent.

File: master/internal/api_runs.go
Comment (AmanuelAaron) [inline master/internal/api_runs.go:734]: nit: "Run is part of archived Search"

File: master/internal/api_runs.go
Comment (AmanuelAaron) [inline master/internal/api_runs.go:732]: I think we want a test for this case in `api_runs_intg_test.go`

File: webui/react/src/utils/flatRun.ts
Comment (keita-determined) [reply webui/react/src/utils/flatRun.ts:9]: i understand the concern. The reason is that permission for runs and experiments are the same for `canCreateExperiment` and `canModifyExperiment`. I don't think it makes sense to make the same functions for runs.

File: master/internal/api_runs.go
Comment (keita-determined) [reply master/internal/api_runs.go]: fixed

**Comment by keita-determined** [comment]
i'll add more test cases. are there anything else i need to change/fix?

File: master/internal/api_runs.go
Comment (keita-determined) [reply master/internal/api_runs.go:732]: i'll add test later today.

Comment (keita-determined) [reply master/internal/api_runs.go:732]: fixed

File: proto/src/determined/run/v1/run.proto
Comment (AmanuelAaron) [inline proto/src/determined/run/v1/run.proto:114]: nit: add that this is only looking at the archived status at the run level and not taking into account whether the experiment is archived or not.

**Review by AmanuelAaron — APPROVED** [review]
BE LGTM. small nit

File: proto/src/determined/run/v1/run.proto
Comment (keita-determined) [reply proto/src/determined/run/v1/run.proto]: fixed

File: webui/react/src/pages/FlatRuns/FlatRunActionButton.tsx
Comment (johnkim-det) [inline webui/react/src/pages/FlatRuns/FlatRunActionButton.tsx:45]: nit: `() => void`

File: webui/react/src/pages/FlatRuns/FlatRuns.tsx
Comment (johnkim-det) [inline webui/react/src/pages/FlatRuns/FlatRuns.tsx:165]: question: does this need to passed down from here? Looks like it's only used in `RunFilterInterstitialModalComponent` so I'm wondering if that component can just import there.

File: webui/react/src/utils/flatRun.ts
Comment (johnkim-det) [reply webui/react/src/utils/flatRun.ts:9]: Personally I think it's more clear to have them use the same name, even if `canCreateFlatRun` and `canModifyFlatRun` just wraps `canCreateExperiment` and `canModifyExperiment` without any changes -- but not something I would block merge on

File: webui/react/src/utils/flatRun.ts
Comment (johnkim-det) [inline webui/react/src/utils/flatRun.ts:31]: nit: already changing this in my PR but, `export`

File: webui/react/src/pages/FlatRuns/FlatRuns.tsx
Comment (keita-determined) [reply webui/react/src/pages/FlatRuns/FlatRuns.tsx:165]: if we export `formStore`, i think this doesnt need to be passed, but passing it down makes writing test easier

File: webui/react/src/pages/FlatRuns/FlatRunActionButton.tsx
Comment (keita-determined) [reply webui/react/src/pages/FlatRuns/FlatRunActionButton.tsx:45]: probably ` () => void | Promise<void>` is more suitable?  `() => void` works for promise, but editor thinks `await` is not needed if `() => void` is used. i think its misleading. what do you think?

File: webui/react/src/utils/flatRun.ts
Comment (keita-determined) [reply webui/react/src/utils/flatRun.ts:31]: fixed

File: webui/react/src/utils/flatRun.ts
Comment (keita-determined) [reply webui/react/src/utils/flatRun.ts]: ok i wrapped these for flatRuns as alias.

File: webui/react/src/pages/FlatRuns/FlatRunActionButton.tsx
Comment (johnkim-det) [reply webui/react/src/pages/FlatRuns/FlatRunActionButton.tsx:45]: sure that sounds good

File: webui/react/src/pages/FlatRuns/FlatRuns.tsx
Comment (johnkim-det) [reply webui/react/src/pages/FlatRuns/FlatRuns.tsx:165]: if testing is the only reason to pass it down and it works fine without it, I'd prefer not passing it down and instead mocking the import in tests as necessary.

Comment (keita-determined) [reply webui/react/src/pages/FlatRuns/FlatRuns.tsx]: fixed

File: webui/react/src/pages/FlatRuns/FlatRunActionButton.tsx
Comment (keita-determined) [reply webui/react/src/pages/FlatRuns/FlatRunActionButton.tsx]: fixed

File: webui/react/src/pages/FlatRuns/FlatRunActionButton.test.tsx
Comment (johnkim-det) [inline webui/react/src/pages/FlatRuns/FlatRunActionButton.test.tsx:52]: nit: 'should not appear',  'should appear'

**Review by johnkim-det — COMMENTED** [review]
I don't have anything else on FE code, but it does look like coverage is not quite 80%. Based on the [report](https://app.codecov.io/gh/determined-ai/determined/pull/9368?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=determined-ai), I *think* covering the submit handlers for `FlatRunMoveModal` and `FlatRunActionButton` should probably be enough?

File: webui/react/src/pages/FlatRuns/FlatRunActionButton.test.tsx
Comment (keita-determined) [reply webui/react/src/pages/FlatRuns/FlatRunActionButton.test.tsx]: fixed

File: webui/react/src/pages/FlatRuns/FlatRunMoveModal.tsx
Comment (johnkim-det) [inline webui/react/src/pages/FlatRuns/FlatRunMoveModal.tsx:132]: should be `Run`

File: webui/react/src/pages/FlatRuns/FlatRuns.tsx
Comment (johnkim-det) [inline webui/react/src/pages/FlatRuns/FlatRuns.tsx:1002]: add selected count?

**Review by johnkim-det — APPROVED** [review]
couple non-blocking things but otherwise LGTM

File: webui/react/src/pages/FlatRuns/FlatRunMoveModal.test.tsx
Comment (johnkim-det) [inline webui/react/src/pages/FlatRuns/FlatRunMoveModal.test.tsx:101]: also `Run` not `Runs`

File: webui/react/src/pages/SearchDetails.test.tsx
Comment (johnkim-det) [inline webui/react/src/pages/SearchDetails.test.tsx:37]: 'Runs' not 'Uncategorized Experiments'

File: webui/react/src/pages/FlatRuns/FlatRuns.tsx
Comment (johnkim-det) [reply webui/react/src/pages/FlatRuns/FlatRuns.tsx:1002]: nvm I guess we created a separate ticket for this

File: webui/react/src/pages/SearchDetails.test.tsx
Comment (keita-determined) [reply webui/react/src/pages/SearchDetails.test.tsx]: fixed

File: webui/react/src/pages/FlatRuns/FlatRunMoveModal.test.tsx
Comment (keita-determined) [reply webui/react/src/pages/FlatRuns/FlatRunMoveModal.test.tsx]: fixed

File: webui/react/src/pages/FlatRuns/FlatRunMoveModal.tsx
Comment (keita-determined) [reply webui/react/src/pages/FlatRuns/FlatRunMoveModal.tsx]: fixed

File: webui/react/src/pages/FlatRuns/FlatRuns.tsx
Comment (keita-determined) [reply webui/react/src/pages/FlatRuns/FlatRuns.tsx:1002]: yep, there's a ticket for it.

**Comment by ashtonG** [comment]
talked to design -- let's use the pencil icon instead of the fire icon for the action menu, please!

**Comment by keita-determined** [comment]
> talked to design -- let's use the pencil icon instead of the fire icon for the action menu, please!

fixed

File: webui/react/src/components/RunMoveWarningModalComponent.tsx
Comment (keita-determined) [inline webui/react/src/components/RunMoveWarningModalComponent.tsx:19]: to avoid "`'` can be escaped with `&apos;`, `&lsquo;`, `&#39;`, `&rsquo;`." error, used string here

File: master/internal/api_runs.go
Comment (corban-beaird) [inline master/internal/api_runs.go:737]: Would we be able to include what search it's part of?

**Review by corban-beaird — APPROVED** [review]
BE side of things looks good, I had one nit about an error message

File: master/internal/api_runs.go
Comment (keita-determined) [reply master/internal/api_runs.go]: fixed it

**Pull request merged**
